pilot/pkg/serviceregistry/kube/controller/serviceexportcontroller.go

```diff
@@ -0,0 +1,159 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"sigs.k8s.io/mcs-api/pkg/apis/v1alpha1"
+	"sigs.k8s.io/mcs-api/pkg/client/clientset/versioned"
+
+	"istio.io/istio/pkg/kube"
+	"istio.io/istio/pkg/queue"
+	"istio.io/pkg/log"
+)
+
+type ServiceExportController struct {
+	client        versioned.Interface
+	serviceClient corev1.CoreV1Interface
+
+	queue           queue.Instance
+	serviceInformer cache.SharedInformer
+
+	clusterLocalHosts []string //hosts marked as cluster-local, which will not have serviceeexports generated
+}
+
+func NewServiceExportController(kubeClient kube.Client, clusterLocalHosts []string) (*ServiceExportController, error) {
+	serviceExportController := &ServiceExportController{
+		client:            kubeClient.MCSApis(),
+		serviceClient:     kubeClient.Kube().CoreV1(),
+		queue:             queue.NewQueue(time.Second),
+		clusterLocalHosts: clusterLocalHosts,
+	}
+
+	serviceExportController.serviceInformer = kubeClient.KubeInformer().Core().V1().Services().Informer()
+	serviceExportController.serviceInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			serviceExportController.queue.Push(func() error {
+				serviceObj, err := convertToService(obj)
+				if err != nil {
+					return err
+				}
+				return serviceExportController.HandleNewService(serviceObj)
+			})
+		},
+		DeleteFunc: func(obj interface{}) {
+			serviceExportController.queue.Push(func() error {
+				serviceObj, err := convertToService(obj)
+				if err != nil {
+					return err
+				}
+				return serviceExportController.HandleDeletedService(serviceObj)
+			})
+		},
+	})
+
+	return serviceExportController, nil
+}
+
+func (sc *ServiceExportController) Run(stopCh <-chan struct{}) {
+	log.Infof("Syncing existing services and serviceexports...")
+	sc.doInitialFullSync()
+	log.Infof("serviceexport sync complete")
+	cache.WaitForCacheSync(stopCh, sc.serviceInformer.HasSynced)
+	log.Infof("ServiceExport controller started")
+	go sc.queue.Run(stopCh)
+}
+
+func (sc *ServiceExportController) HandleNewService(obj *v1.Service) error {
+	if sc.isServiceClusterLocal(obj) {
+		return nil //don't don anything for marked clusterlocal services
+	}
+	return sc.createServiceExportIfNotPresent(obj)
+}
+
+func (sc *ServiceExportController) HandleDeletedService(obj *v1.Service) error {
+	return sc.deleteServiceExportIfPresent(obj)
+}
+
+func convertToService(obj interface{}) (*v1.Service, error) {
+	cm, ok := obj.(*v1.Service)
+	if !ok {
+		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			return nil, fmt.Errorf("couldn't get object from tombstone %#v", obj)
+		}
+		cm, ok = tombstone.Obj.(*v1.Service)
+		if !ok {
+			return nil, fmt.Errorf("tombstone contained object that is not a Service %#v", obj)
+		}
+	}
+	return cm, nil
+}
+
+func (sc *ServiceExportController) isServiceClusterLocal(service *v1.Service) bool {
```

Comment:
Can we move this function to the context?

Reply:
I'm not sure what you mean by this, can you clarify?

Reply:
The context is where you're getting information regarding cluster-local services. Could it just expose this method instead of defining it here?